File: Mediator/java/Mediator.java
interface IMediator{
    void notify(Employee employee, String msg);
}

abstract class Employee{
    protected IMediator mediator;

    public Employee(IMediator mediator){
        this.mediator = mediator;
    }
    
    public void setMediator(IMediator mediator){
        this.mediator = mediator;
    }
}

class Designer extends Employee{
    private boolean isWorking;
    
    public Designer(IMediator mediator){
        super(mediator);
    }
    
    public void executeWork(){
        System.out.println("<-the designer is at work");
        mediator.notify(this, "the designer designs...");
    }
    
    public void setWork(boolean state){
        isWorking = state;
        if (state)
            System.out.println("<-the designer is released from work");
        else
            System.out.println("<-the designer is busy");
    }
}

class Director extends Employee{
    private String text;
    
    public Director(IMediator mediator){
        super(mediator);
    }
    
    public void giveCommand(String txt){
        text = txt;
        if (text == "")
            System.out.println("->the director knows that the designer is already working");
        else
            System.out.println("->the director gave the command:" + text);
        mediator.notify(this, text);
    }
}

class Controller implements IMediator{
    private Designer designer;
    private Director director;
    
    public Controller(Designer designer, Director director){
        this.designer = designer;
        this.director = director;
        designer.setMediator(this);
        director.setMediator(this);
    }
    
    @Override
    public void notify(Employee employee, String msg){
        if (employee instanceof Director) {
            if (msg.equals(""))
                designer.setWork(false);
            else
                designer.setWork(true);
        }
        if (employee instanceof Designer)
            director.giveCommand("");
    }
}

public class Main {
    public static void main(String[] args){
        Designer designer = new Designer(null);
        Director director = new Director(null);

        Controller mediator = new Controller(designer, director);

        director.giveCommand("design");

        System.out.println();

        designer.executeWork();
    }
}

File: Mediator/python/Mediator.py
import abc


class IMediator(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def notify(self, emp: 'Employee', msg: str):
        pass


class Employee(metaclass=abc.ABCMeta):

    def __init__(self, mediator: IMediator):
        self._mediator = mediator

    def set_mediator(self, med: IMediator):
        self._mediator = med


class Designer(Employee):

    def __init__(self, med: IMediator = None):
        super().__init__(med)
        self.__is_working = False

    def execute_work(self):
        print("<-the designer is at work")
        self._mediator.notify(self, "the designer designs...")

    def set_work(self, state: bool):
        self.__is_working = state
        if state:
            print("<-the designer is released from work")
        else:
            print("<-the designer is busy")


class Director(Employee):

    def __init__(self, med: IMediator = None):
        super().__init__(med)
        self.__text: str = None

    def give_command(self, txt: str):
        self.__text = txt
        if txt == "":
            print("->the director knows that the designer is already working")
        else:
            print("->the director gave the command:" + txt)
        self._mediator.notify(self, txt)


class Controller(IMediator):

    def __init__(self, designer: Designer, director: Director):
        self.__designer = designer
        self.__director = director
        designer.set_mediator(self)
        director.set_mediator(self)

    def notify(self, emp: 'Employee', msg: str):
        if isinstance(emp, Director):
            if msg == "":
                self.__designer.set_work(False)
            else:
                self.__designer.set_work(True)

        if isinstance(emp, Designer):
            if msg == "the designer designs...":
                self.__director.give_command("")


if __name__ == '__main__':
    designer = Designer()
    director = Director()

    mediator = Controller(designer, director)

    director.give_command("design")

    print()

    designer.execute_work()